Source/WebCore/svg/SVGLengthValue.cpp

```diff
@@ -350,35 +430,59 @@ ExceptionOr<void> SVGLengthValue::setValueAsString(StringView string)
     if (string.isEmpty())
         return { };
 
-    return readCharactersForParsing(string, [&](auto buffer) -> ExceptionOr<void> {
-        auto convertedNumber = parseNumber(buffer, SuffixSkippingPolicy::DontSkip);
-        if (!convertedNumber)
+    // FIXME: SVG2 allows trailing spaces?
+    if (isASCIIWhitespace(string[string.length() - 1]))
+        return Exception { ExceptionCode::SyntaxError };
+
+    auto parserContext = CSSParserContext { SVGAttributeMode };
+    auto parserState = CSS::PropertyParserState {
+        .context = parserContext
+    };
+
+    CSSTokenizer tokenizer(string.toString());
+    auto tokenRange = tokenizer.tokenRange();
+
+    if (auto number = CSSPropertyParserHelpers::MetaConsumer<CSS::Number<>>::consume(tokenRange, parserState, { })) {
+        if (!tokenRange.atEnd())
             return Exception { ExceptionCode::SyntaxError };
 
-        auto lengthType = parseLengthType(buffer);
-        if (lengthType == SVGLengthType::Unknown)
+        m_value = WTFMove(*number);
+        return { };
+    }
+
+    tokenRange = tokenizer.tokenRange();
+    if (auto length = CSSPropertyParserHelpers::MetaConsumer<CSS::LengthPercentage<>>::consume(tokenRange, parserState, { })) {
+        if (!tokenRange.atEnd())
+            return Exception { ExceptionCode::SyntaxError };
+
+        // FIXME: Add support for calc
+        // For now, reject it
+        if (length->isCalc())
             return Exception { ExceptionCode::SyntaxError };
 
-        m_lengthType = lengthType;
-        m_valueInSpecifiedUnits = *convertedNumber;
+        m_value = WTFMove(*length);
         return { };
-    });
+    }
+
+    return Exception { ExceptionCode::SyntaxError };
 }
 
-ExceptionOr<void> SVGLengthValue::convertToSpecifiedUnits(const SVGLengthContext& context, SVGLengthType lengthType)
+ExceptionOr<void> SVGLengthValue::convertToSpecifiedUnits(const SVGLengthContext& context, SVGLengthType targetType)
 {
     auto valueInUserUnits = valueForBindings(context);
     if (valueInUserUnits.hasException())
         return valueInUserUnits.releaseException();
 
-    auto originalLengthType = m_lengthType;
-    m_lengthType = lengthType;
-    auto result = setValue(context, valueInUserUnits.releaseReturnValue());
-    if (!result.hasException())
-        return { };
+    auto convertedValue = context.convertValueFromUserUnits(valueInUserUnits.releaseReturnValue(), targetType, m_lengthMode);
 
-    m_lengthType = originalLengthType;
-    return result.releaseException();
+    if (convertedValue.hasException())
+        return convertedValue.releaseException();
+
+    auto adjustedValue = adjustValueForPercentageStorage(convertedValue.releaseReturnValue(), targetType);
+
+    m_value = createVariantForLengthType(adjustedValue, targetType);
```

Comment:
Would adding a move be helpful?
```suggestion
    m_value = createVariantForLengthType(WTFMove(adjustedValue), targetType);
```

Reply:
`adjustedValue` is a primitive type, not an expensive object. perhaps, i should make that clearer. i'll replace auto with float so it's more evident